File: svc-data/src/test/java/com/beancounter/marketdata/integ/TestRegistrationController.java
package com.beancounter.marketdata.integ;

import static org.assertj.core.api.Assertions.assertThat;
import static org.springframework.security.test.web.servlet.request.SecurityMockMvcRequestPostProcessors.csrf;
import static org.springframework.security.test.web.servlet.request.SecurityMockMvcRequestPostProcessors.jwt;
import static org.springframework.security.test.web.servlet.setup.SecurityMockMvcConfigurers.springSecurity;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

import com.beancounter.auth.common.TokenUtils;
import com.beancounter.auth.server.AuthorityRoleConverter;
import com.beancounter.common.model.SystemUser;
import com.beancounter.marketdata.utils.RegistrationUtils;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Tag;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;
import org.springframework.security.oauth2.jwt.Jwt;
import org.springframework.test.context.junit.jupiter.SpringExtension;
import org.springframework.test.web.servlet.MockMvc;
import org.springframework.test.web.servlet.MvcResult;
import org.springframework.test.web.servlet.setup.MockMvcBuilders;
import org.springframework.web.context.WebApplicationContext;

@ExtendWith(SpringExtension.class)
@SpringBootTest
@Tag("slow")
public class TestRegistrationController {
  private MockMvc mockMvc;

  @Autowired
  private WebApplicationContext context;

  @BeforeEach
  void mockServices() {
    this.mockMvc = MockMvcBuilders.webAppContextSetup(context)
        .apply(springSecurity())
        .build();
  }

  @Test
  void is_RegisterMeWorking() throws Exception {
    SystemUser user = new SystemUser("user", "user@testing.com");
    Jwt token = TokenUtils.getUserToken(user);
    RegistrationUtils.registerUser(mockMvc, token);

    MvcResult performed = mockMvc.perform(
        get("/me")
            .with(jwt().jwt(token).authorities(new AuthorityRoleConverter()))
            .with(csrf())
            .contentType(MediaType.APPLICATION_JSON)
    ).andExpect(status().is2xxSuccessful())
        .andReturn();
    assertThat(performed.getResponse().getStatus()).isEqualTo(HttpStatus.OK.value());
  }

  @Test
  void is_MeWithNoToken() throws Exception {
    SystemUser user = new SystemUser("user", "user@testing.com");
    Jwt token = TokenUtils.getUserToken(user);
    RegistrationUtils.registerUser(mockMvc, token);

    MvcResult performed = mockMvc.perform(
        get("/me")
            .with(csrf())
            .contentType(MediaType.APPLICATION_JSON)
    ).andExpect(status().is4xxClientError())
        .andReturn();

    assertThat(performed.getResponse().getStatus()).isEqualTo(HttpStatus.UNAUTHORIZED.value());
  }

  @Test
  void is_MeUnregistered() throws Exception {
    SystemUser user = new SystemUser(
        "is_MeUnregistered",
        "is_MeUnregistered@testing.com");
    Jwt token = TokenUtils.getUserToken(user);

    MvcResult performed = mockMvc.perform(
        get("/me")
            .with(jwt().jwt(token).authorities(new AuthorityRoleConverter()))
            .with(csrf())
            .contentType(MediaType.APPLICATION_JSON)
    ).andExpect(status().is4xxClientError())
        .andReturn();

    assertThat(performed.getResponse().getStatus()).isEqualTo(HttpStatus.FORBIDDEN.value());
  }

}

File: svc-data/src/test/java/com/beancounter/marketdata/integ/TestRegistrationController.kt
package com.beancounter.marketdata.integ

import com.beancounter.auth.common.TokenUtils
import com.beancounter.auth.server.AuthorityRoleConverter
import com.beancounter.common.model.SystemUser
import com.beancounter.marketdata.utils.RegistrationUtils.registerUser
import org.assertj.core.api.Assertions
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Tag
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.autoconfigure.domain.EntityScan
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.http.HttpStatus
import org.springframework.http.MediaType
import org.springframework.security.test.web.servlet.request.SecurityMockMvcRequestPostProcessors
import org.springframework.security.test.web.servlet.setup.SecurityMockMvcConfigurers
import org.springframework.test.context.ActiveProfiles
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders
import org.springframework.test.web.servlet.result.MockMvcResultMatchers
import org.springframework.test.web.servlet.setup.DefaultMockMvcBuilder
import org.springframework.test.web.servlet.setup.MockMvcBuilders
import org.springframework.web.context.WebApplicationContext

@SpringBootTest
@ActiveProfiles("test")
@Tag("slow")
@EntityScan("com.beancounter.common.model")
class TestRegistrationController {
    private lateinit var mockMvc: MockMvc
    private val tokenUtils = TokenUtils()

    @Autowired
    private lateinit var context: WebApplicationContext

    @BeforeEach
    fun mockServices() {
        mockMvc = MockMvcBuilders.webAppContextSetup(context)
            .apply<DefaultMockMvcBuilder>(SecurityMockMvcConfigurers.springSecurity())
            .build()
    }

    @Test
    @Throws(Exception::class)
    fun is_RegisterMeWorking() {
        val user = SystemUser("user", "user@testing.com")
        val token = tokenUtils.getUserToken(user)
        registerUser(mockMvc, token)
        val performed = mockMvc.perform(
            MockMvcRequestBuilders.get("/me")
                .with(SecurityMockMvcRequestPostProcessors.jwt().jwt(token).authorities(AuthorityRoleConverter()))
                .with(SecurityMockMvcRequestPostProcessors.csrf())
                .contentType(MediaType.APPLICATION_JSON)
        ).andExpect(MockMvcResultMatchers.status().is2xxSuccessful)
            .andReturn()
        Assertions.assertThat(performed.response.status).isEqualTo(HttpStatus.OK.value())
    }

    @Test
    @Throws(Exception::class)
    fun is_MeWithNoToken() {
        val user = SystemUser("user", "user@testing.com")
        val token = tokenUtils.getUserToken(user)
        registerUser(mockMvc, token)
        val performed = mockMvc.perform(
            MockMvcRequestBuilders.get("/me")
                .with(SecurityMockMvcRequestPostProcessors.csrf())
                .contentType(MediaType.APPLICATION_JSON)
        ).andExpect(MockMvcResultMatchers.status().is4xxClientError)
            .andReturn()
        Assertions.assertThat(performed.response.status).isEqualTo(HttpStatus.UNAUTHORIZED.value())
    }

    @Test
    @Throws(Exception::class)
    fun is_MeUnregistered() {
        val user = SystemUser(
            "is_MeUnregistered",
            "is_MeUnregistered@testing.com"
        )
        val token = tokenUtils.getUserToken(user)
        val performed = mockMvc.perform(
            MockMvcRequestBuilders.get("/me")
                .with(SecurityMockMvcRequestPostProcessors.jwt().jwt(token).authorities(AuthorityRoleConverter()))
                .with(SecurityMockMvcRequestPostProcessors.csrf())
                .contentType(MediaType.APPLICATION_JSON)
        ).andExpect(MockMvcResultMatchers.status().is4xxClientError)
            .andReturn()
        Assertions.assertThat(performed.response.status).isEqualTo(HttpStatus.FORBIDDEN.value())
    }
}
